plugins/ua_nodestore_binary_encoded.c

```diff
@@ -0,0 +1,1163 @@
+        variableNode->arrayDimensions = (UA_UInt32 *)UA_calloc(variableNode->arrayDimensionsSize, sizeof(UA_UInt32));
+        retval |= UA_UInt32_decodeBinary(src, offset, variableNode->arrayDimensions);
+    }
+    retval |= UA_UInt32_decodeBinary(src, offset, (UA_UInt32*)&variableNode->valueSource);
+    retval |= UA_DataValue_decodeBinary(src, offset, &variableNode->value.data.value);
+    return retval;
+}
+
+static UA_StatusCode
+methodNodeDecode(const UA_ByteString *src, size_t *offset, UA_MethodNode* methodNode) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    retval |= UA_Boolean_decodeBinary(src, offset, &methodNode->executable);
+    return retval;
+}
+
+static UA_StatusCode
+objectTypeNodeDecode(const UA_ByteString *src, size_t *offset, UA_ObjectTypeNode* objTypeNode) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    retval |= UA_Boolean_decodeBinary(src, offset, &objTypeNode->isAbstract);
+    return retval;
+}
+
+static UA_StatusCode
+variableTypeNodeDecode(const UA_ByteString *src, size_t *offset, UA_VariableTypeNode* varTypeNode) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    retval |= UA_Boolean_decodeBinary(src, offset, &varTypeNode->isAbstract);
+    retval |= UA_NodeId_decodeBinary(src, offset, &varTypeNode->dataType);
+    retval |= UA_Int32_decodeBinary(src, offset, &varTypeNode->valueRank);
+    retval |= UA_UInt64_decodeBinary(src, offset, &varTypeNode->arrayDimensionsSize);
+    if(varTypeNode->arrayDimensionsSize) {
+        retval |= UA_UInt32_decodeBinary(src, offset, &varTypeNode->arrayDimensions[0]);
+    }
+    retval |= UA_UInt32_decodeBinary(src, offset, (UA_UInt32*)&varTypeNode->valueSource);
+    retval |= UA_DataValue_decodeBinary(src, offset, &varTypeNode->value.data.value);
+
+    return retval;
+}
+
+static UA_StatusCode
+ReferenceTypeNodeDecode(const UA_ByteString *src, size_t *offset, UA_ReferenceTypeNode* refTypeNode) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    retval |= UA_Boolean_decodeBinary(src, offset, &refTypeNode->isAbstract);
+    retval |= UA_Boolean_decodeBinary(src, offset, &refTypeNode->symmetric);
+    retval |= UA_LocalizedText_decodeBinary(src, offset, &refTypeNode->inverseName);
+    return retval;
+}
+
+static UA_StatusCode
+dataTypeNodeDecode(const UA_ByteString *src, size_t *offset, UA_DataTypeNode* dataTypeNode) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    retval |= UA_Boolean_decodeBinary(src, offset, &dataTypeNode->isAbstract);
+    return retval;
+}
+
+static UA_StatusCode
+viewNodeDecode(const UA_ByteString *src, size_t *offset, UA_ViewNode* viewNode) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    retval |= UA_Byte_decodeBinary(src, offset, &viewNode->eventNotifier);
+    retval |= UA_Boolean_decodeBinary(src, offset, &viewNode->containsNoLoops);
+    return retval;
+}
+
+
+
+static void
+readRowLookuptable(char ltRow [], int length, lookUpTable *lt, int row) {
+    /**
+     * There are four contents per row
```

Comment:
There is a standard text format for NodeIds.
Like `i=13`, `ns=10;s=Hello:World`, `g=09087e75-8e5e-499b-954f-f2a9603db28a`, `ns=1;b=M/RbKBsRVkePCePcx24oRA==`, etc.
See Part 6, Section 5.3.1.10. Let's try to write and parse that.

So we are known to be complete.

Base64 encoding/decoding is included in the dependencies.
And already used for the JSON parser.

Here's a snippet to parse guids of the form `00000000-0000-0000-0000-000000000000` from a char-array `arg`:
```c
    UA_Guid *guid = UA_Guid_new();
    guid->data1 = strtoull(arg, NULL, 16);
    guid->data2 = strtoull(&arg[9], NULL, 16);
    guid->data3 = strtoull(&arg[14], NULL, 16);
    UA_Int16 data4_1 = strtoull(&arg[19], NULL, 16);
    guid->data4[0] = data4_1 >> 8;
    guid->data4[1] = data4_1;
    UA_Int64 data4_2 = strtoull(&arg[24], NULL, 16);
    guid->data4[2] = data4_2 >> 40;
    guid->data4[3] = data4_2 >> 32;
    guid->data4[4] = data4_2 >> 24;
    guid->data4[5] = data4_2 >> 16;
    guid->data4[6] = data4_2 >> 8;
    guid->data4[7] = data4_2;
```

Reply:
Here's an extended snippet for NodeId parsing I still had from an older project.
You can quote form that.

```c
/* Parse NodeId from String */
static UA_StatusCode
parseNodeId(UA_String str, UA_NodeId *id) {
                /* zero-terminated string */
    UA_STACKARRAY(char, s, str.length + 1);
                memcpy(s, str.data, str.length);
                s[str.length] = 0;

                UA_NodeId_init(id);

                /* Int identifier */
                if(sscanf(s, "ns=%hu;i=%u", &id->namespaceIndex, &id->identifier.numeric) == 2 ||
       sscanf(s, "ns=%hu,i=%u", &id->namespaceIndex, &id->identifier.numeric) == 2 ||
       sscanf(s, "i=%u", &id->identifier.numeric) == 1) {
                               id->identifierType = UA_NODEIDTYPE_NUMERIC;
                               return UA_STATUSCODE_GOOD;
                }

    /* String identifier */
    UA_STACKARRAY(char, sid, str.length);
                if(sscanf(s, "ns=%hu;s=%[^\n]", &id->namespaceIndex, sid) == 2 ||
       sscanf(s, "ns=%hu,s=%[^\n]", &id->namespaceIndex, sid) == 2 ||
       sscanf(s, "s=%s", sid) == 1) {
                               id->identifierType = UA_NODEIDTYPE_STRING;
                               id->identifier.string = UA_String_fromChars(sid);
                               return UA_STATUSCODE_GOOD;
    }

    /* ByteString idenifier */
                if(sscanf(s, "ns=%hu;b=%[^\n]", &id->namespaceIndex, sid) == 2 ||
       sscanf(s, "ns=%hu,b=%[^\n]", &id->namespaceIndex, sid) == 2 ||
       sscanf(s, "b=%s", sid) == 1) {
                               id->identifierType = UA_NODEIDTYPE_BYTESTRING;
                               id->identifier.string = UA_String_fromChars(sid);
                               return UA_STATUSCODE_GOOD;
    }

    /* Guid idenifier */
                if(sscanf(s, "ns=%hu;g=%s", &id->namespaceIndex, sid) == 2 ||
       sscanf(s, "ns=%hu,g=%s", &id->namespaceIndex, sid) == 2 ||
       sscanf(s, "g=%s", sid) == 1) {
        UA_StatusCode retval = parseGuid(sid, &id->identifier.guid);
        if(retval != UA_STATUSCODE_GOOD)
            return retval;
                               id->identifierType = UA_NODEIDTYPE_GUID;
                               return UA_STATUSCODE_GOOD;
    }

                return UA_STATUSCODE_BADINTERNALERROR;
}
```